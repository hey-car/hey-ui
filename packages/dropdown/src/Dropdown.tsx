/* eslint-disable prettier/prettier */
import React, { useState } from 'react';
import cn from 'classnames';

import { DropdownProps, SelectOptions } from './Dropdown.types';

import styles from './styles/default.module.css';

export const Dropdown = ({
  value,
  onChange,
  options,
  disabled,
  dataTestId,
  onBlur,
  onClick,
  fullWidth,
}: DropdownProps) => {
  const [stateValue, setStateValue] = useState<SelectOptions | undefined>(
    value,
  );

  const [isOpen, setIsOpen] = useState(false);

  const selectOption = (option: SelectOptions) => {
    if (onChange) {
      if (option !== stateValue) onChange(option);
    }
    setStateValue(option);
  };

  const isOptionSelection = (option: SelectOptions) => {
    return option?.value === stateValue?.value;
  };

  const onClickHandler = () => {
    if (onClick) onClick();
    if (!disabled) setIsOpen(!isOpen);
  };
  const onBlurHandler = () => {
    if (onBlur) onBlur();
    setIsOpen(false);
  };

  const classNames = cn(
    styles.container,
    disabled && styles.disabled,
    fullWidth && styles.fullWidth,
  );

  const arrowClassNames = cn(
    disabled && styles.caret_disabled,
    isOpen && styles.caret_up,
    !isOpen && styles.caret_down,
  );

  const valueClassNames = cn(
    disabled && 'disabled',
    styles.value,
  );

  return (
    <div
      className={classNames}
      onBlur={onBlurHandler}
      onClick={onClickHandler}
      tabIndex={0}
    >
      <span className={valueClassNames}>
        {stateValue?.label}
      </span>
      <div className={arrowClassNames}></div>
      <ul
        className={`${styles.options} ${isOpen ? styles.show : ''}`}
        data-test-id={dataTestId}
      >
        {options.map(option => (
          <li
            className={`${styles.option} ${isOptionSelection(option) ? styles.selected : ''
              }`}
            key={option.value}
            onClick={e => {
              e.stopPropagation();
              selectOption(option);
              setIsOpen(false);
            }}
          >
            <div className={styles.contentContainer}>
              {option.leftContent && (
                <span className={styles.leftContent}>{option.leftContent}</span>
              )}
              <span className={styles.optionContent}>{option.label}</span>
              {option.rightContent && (
                <span className={styles.rightContent}>
                  {option.rightContent}
                </span>
              )}
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Dropdown;
