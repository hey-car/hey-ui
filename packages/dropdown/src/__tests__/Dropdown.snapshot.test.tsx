import React from 'react';
import { render } from '@testing-library/react';

import Dropdown from '../Dropdown';

describe('Dropdown Snapshots tests', () => {
  it('should match callback methods', () => {
    expect(
      render(
        <Dropdown
          onChange={jest.fn()}
          options={[
            {
              value: 'pomelo',
              label: 'Pomelo',
            },
            {
              value: 'apple',
              label: 'Apple',
            },
            {
              value: 'mango',
              label: 'Mango',
            },
          ]}
          value={{
            value: 'mango',
            label: 'Mango',
          }}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should match disabled state', () => {
    expect(
      <Dropdown
        disabled={true}
        onChange={() => {}}
        options={[
          {
            value: 'pomelo',
            label: 'Pomelo',
          },
          {
            value: 'apple',
            label: 'Apple',
          },
          {
            value: 'mango',
            label: 'Mango',
          },
        ]}
        value={{
          value: 'mango',
          label: 'Mango',
        }}
      />,
    ).toMatchSnapshot();
  });
});
