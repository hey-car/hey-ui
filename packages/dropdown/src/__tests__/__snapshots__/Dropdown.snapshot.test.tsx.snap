// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Snapshots tests should match callback methods 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        tabindex="0"
      >
        <span
          class="value"
        >
          Mango
        </span>
        <div
          class="caret_down"
        />
        <ul
          class="options "
        >
          <li
            class="option "
          >
            <div
              class="contentContainer"
            >
              <span
                class="optionContent"
              >
                Pomelo
              </span>
            </div>
          </li>
          <li
            class="option "
          >
            <div
              class="contentContainer"
            >
              <span
                class="optionContent"
              >
                Apple
              </span>
            </div>
          </li>
          <li
            class="option selected"
          >
            <div
              class="contentContainer"
            >
              <span
                class="optionContent"
              >
                Mango
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      tabindex="0"
    >
      <span
        class="value"
      >
        Mango
      </span>
      <div
        class="caret_down"
      />
      <ul
        class="options "
      >
        <li
          class="option "
        >
          <div
            class="contentContainer"
          >
            <span
              class="optionContent"
            >
              Pomelo
            </span>
          </div>
        </li>
        <li
          class="option "
        >
          <div
            class="contentContainer"
          >
            <span
              class="optionContent"
            >
              Apple
            </span>
          </div>
        </li>
        <li
          class="option selected"
        >
          <div
            class="contentContainer"
          >
            <span
              class="optionContent"
            >
              Mango
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dropdown Snapshots tests should match disabled state 1`] = `
<Dropdown
  disabled={true}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "Pomelo",
        "value": "pomelo",
      },
      Object {
        "label": "Apple",
        "value": "apple",
      },
      Object {
        "label": "Mango",
        "value": "mango",
      },
    ]
  }
  value={
    Object {
      "label": "Mango",
      "value": "mango",
    }
  }
/>
`;
